File: CourseSchedule.py
class Solution:
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = [ [] for _ in range(numCourses)]
        for pre in prerequisites:
            graph[pre[1]].append(pre[0])
        indegree = [0] * numCourses
        for neighbors in graph:
            for neighbor in neighbors:
                indegree[neighbor] += 1
        queue = []
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        count = 0
        while len(queue) > 0:
            t = queue[0]
            queue.pop(0)
            count += 1
            for neighbor in graph[t]:
                indegree[neighbor] -= 1
                if indegree[neighbor] == 0:
                    queue.append(neighbor)
        return count == numCourses
        
if __name__ == '__main__':
    numCourses = int(raw_input())
    n = int(raw_input())
    prerequisites = []
    for i in range(n):
        pre = list(input())
        prerequisites.append(pre)
    solution = Solution()
    print(solution.canFinish(numCourses, prerequisites))

File: CourseSchedule.cpp
#include<iostream>
#include<vector>
#include<queue>

using namespace std;

class Solution {
public:
    bool canFinish(int numCourses, vector<pair<int, int> >& prerequisites) {
        vector<vector<int> > graph(numCourses, vector<int>(numCourses, 0));
        vector<int> in(numCourses, 0);
        
        for(int i=0;i<prerequisites.size();i++){
            graph[prerequisites[i].second][prerequisites[i].first] = 1;
            in[prerequisites[i].first] ++;
        }
        
        queue<int> nodes;
        for(int i=0;i<numCourses;i++){
            if(in[i] == 0) nodes.push(i);
        }
        
        /*while(!nodes.empty()) {
            int t = nodes.front();
            nodes.pop();
            for(int i=0;i<numCourses;i++){
                if(graph[t][i]){
                    in[i] --;
                    if(in[i] == 0) nodes.push(i);
                }
            }
        }
        
        for(int i=0;i<numCourses;i++){
            if(in[i] != 0) return false;
        }
        return true; 
        */
        
        //另一种方式
        int count = 0;
        while(!nodes.empty()) {
            int t = nodes.front();
            nodes.pop();
            count ++;
            for(int i=0;i<numCourses;i++){
                if(graph[t][i]){
                    in[i] --;
                    if(in[i] == 0) nodes.push(i);
                }
            }
        }
        return count == numCourses;
        
    }
};

int main(){
    
    int numCourses;
    cin>>numCourses;
    
    int n;
    cin>>n;
    vector<pair<int, int> > prerequisites;
    for(int i=0;i<n;i++){
        int a, b;
        cin>>a>>b;
        prerequisites.push_back(pair<int, int>(a, b));
    }
    
    Solution *solution = new Solution();
    solution->canFinish(numCourses, prerequisites) ? cout<<"True" : cout<<"False";
    
    return 0;
    
}
